src/smc-project/sync/usage-info.ts

```diff
@@ -0,0 +1,184 @@
+/*
+ *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
+ *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
+ */
+
+// usage info for a specific file path, derived from the more general project info,
+// which includes all processes and other stats
+
+import * as debug from "debug";
+const L = debug("project:sync:usage-info");
+import { once } from "../smc-util/async-utils";
+import { SyncTable, SyncTableState } from "../smc-util/sync/table";
+import { close, merge } from "../smc-util/misc";
+import { UsageInfoServer } from "../usage-info";
+import { UsageInfo, ImmutableUsageInfo } from "../usage-info/types";
+
+class UsageInfoTable {
+  private readonly table: SyncTable;
+  private readonly project_id: string;
+  private readonly servers: { [path: string]: UsageInfoServer } = {};
+  private readonly log: Function;
+
+  constructor(table: SyncTable, project_id: string) {
+    this.project_id = project_id;
+    this.log = L.extend("table");
+    this.table = table;
+    this.setup_watchers();
+  }
+
+  public close(): void {
+    this.log("close");
+    for (const path in this.servers) {
+      this.stop_server(path);
+    }
+    close(this);
+  }
+
+  // Start watching any paths that have recent interest (so this is not
+  // in response to a *change* after starting).
+  private async setup_watchers(): Promise<void> {
+    if (this.table == null) return; // closed
+    if (this.table.get_state() == ("init" as SyncTableState)) {
+      await once(this.table, "state");
+    }
+    if (this.table.get_state() != ("connected" as SyncTableState)) {
+      return; // game over
+    }
+    this.table.get()?.forEach((val) => {
+      const path = val.get("path");
+      if (path == null) return;
+      if (this.servers[path] == null) return; // already watching
+    });
+    this.log("setting up 'on.change'");
+    this.table.on("change", this.handle_change_event.bind(this));
+  }
+
+  private async remove_stale_servers(): Promise<void> {
+    if (this.table == null) return; // closed
+    if (this.table.get_state() != ("connected" as SyncTableState)) return;
+    const paths: string[] = [];
+    this.table.get()?.forEach((val) => {
+      const path = val.get("path");
+      if (path == null) return;
+      paths.push(path);
+    });
+    for (const path of Object.keys(this.servers)) {
+      if (!paths.includes(path)) {
+        this.stop_server(path);
+      }
+    }
+  }
+
+  private is_ready(): boolean {
+    return this.table?.is_ready();
```

Comment:
this is all copy/paste. does this mean it's also not ok for listings? i.e. this method: https://github.com/sagemathinc/cocalc/blob/master/src/smc-project/sync/listings.ts#L149

same for table there.

or is there something else different in listings that isn't here?

Reply:
I checked this, I will actually remove the  question mark. There is `sync/server.ts` and well, there it gets this `synctable` from `SyncTableChannel` and there is no way it will be removed.

Reply:
oh well, there is also this funny close method. I assume that's what this is about.